### PR TITLE
[SHLWAPI] Implement SHLockSharedEx function

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
- Implement `shlwapi!SHLockSharedEx` function.
- Modify `shlwapi.spec` file (some argument conversions from `long` to `ptr`).

JIRA issue: [CORE-13950](https://jira.reactos.org/browse/CORE-13950)

### CODE REVIEW

**Comment by SergeGautherie** [comment]
To send to upstream Wine?

**Comment by katahiromz** [comment]
I wanna send this to Wine.

**Comment by HBelusca** [comment]
If you want to send this to Wine, I would then suggest you that for implementing SHLockSharedEx you just 100%-copy the code of SHLockShared, keeping the **exact same local variable names** but add the dwAccess support as you did. Then SHLockShared just calls SHLockSharedEx as you did.

File: dll/win32/shlwapi/ordinal.c [hunk 1]
```diff
@@ -212,25 +213,23 @@ HANDLE WINAPI SHAllocShared(LPCVOID lpvData, DWORD dwSize, DWORD dwProcId)
...
+  HANDLE hDup;
+  LPVOID pMapped;
+  DWORD dwAccess;
 
-    TRACE("(%p %d %d)\n", hShared, dwProcId, bWriteAccess);
+  TRACE("(%p %d %d)\n", hShared, dwProcId, bWriteAccess);
 
-    /* Get handle to shared memory for current process */
-    hDup = SHMapHandle(hShared, dwProcId, GetCurrentProcessId(), FILE_MAP_ALL_ACCESS, 0);
-    if (!hDup)
```
Comment (SergeGautherie) [inline dll/win32/shlwapi/ordinal.c]: Why remove this check? It looked like an improvement...

Comment (HBelusca) [reply dll/win32/shlwapi/ordinal.c]: I would then suggest @katahiromz to make a separate PR / send a separate patch to wine for this.

Comment (katahiromz) [reply dll/win32/shlwapi/ordinal.c]: If hDup is NULL, then MapViewOfFile will return NULL.

**Comment by katahiromz** [comment]
https://bugs.winehq.org/show_bug.cgi?id=48736

File: dll/win32/shlwapi/ordinal.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shlwapi/ordinal.c]: Fixed in 5a440b2.

Comment (SergeGautherie) [reply dll/win32/shlwapi/ordinal.c]: > separate

(Agreed.)

> MapViewOfFile will return NULL

From a Wine point of view, they may prefer less checks.

> Fixed

From a ReactOS point of view, I would suggest to add `if (hShared == NULL) return NULL;` even, assuming that matches Windows result.

Comment (SergeGautherie) [reply dll/win32/shlwapi/ordinal.c]: > 5a440b2

You updated the wrong function.

An additional tweak: `dwProcId` should have `%u`.

**Comment by SergeGautherie** [comment]
> https://bugs.winehq.org/show_bug.cgi?id=48736

Filing a bug is probably fine, though I assume you want to actually submit the patch(es):
https://wiki.winehq.org/Submitting_Patches

**Pull request merged**